### ISSUE TITLE
bugs in CreateOrleansTables_MySql.sql - creates concurrency issues

### ISSUE THREAD

**Issue by shayhatsor** [dotnet/orleans#1169]
During our testing we've came up with a periodical MySql deadlock error while trying to upsert a reminder. the problem seems to manifest because:
* the [SELECT... FOR UPDATE] (http://www.techonthenet.com/oracle/cursors/for_update.php) statement doesn't lock a row that doesn't exist yet.
* Using a grain timer. It is reentrant so we concurrently tried to upsert the same grain reminder.

I'm working on a solution for all similar cases in the script. just wanted to put it here so people like @veikkoeeva can help.
a dump of one of the errors:
```sql
LATEST DETECTED DEADLOCK
------------------------
2015-12-16 17:21:58 7f3b33a4d700
*** (1) TRANSACTION:
TRANSACTION 9133558, ACTIVE 0 sec inserting
mysql tables in use 1, locked 1
LOCK WAIT 3 lock struct(s), heap size 1184, 2 row lock(s)
MySQL thread id 30233, OS thread handle 0x7f3b33804700, query id 42183680 10.201.200.37 orleans_user update
INSERT INTO OrleansRemindersTable
			(
				ServiceId,
				GrainId,
				ReminderName,
				StartTime,
				Period,
				GrainIdConsistentHash
			)
			VALUES
			(
				_serviceId,
				_grainId,
				_reminderName,
				_startTime,
				_period,
				_grainIdConsistentHash
			)
*** (1) WAITING FOR THIS LOCK TO BE GRANTED:
RECORD LOCKS space id 1082 page no 3 n bits 168 index `PRIMARY` of table `orleans`.`OrleansRemindersTable` trx id 9133558 lock_mode X locks gap before rec insert intention waiting
Record lock, heap no 70 PHYSICAL RECORD: n_fields 9; compact format; info bits 0
 0: len 30; hex 30303030303030302d303030302d303030302d303030302d303030306235; asc 00000000-0000-0000-0000-0000b5; (total 36 bytes);
 1: len 30; hex 477261696e5265666572656e63653d656530383038383130373338626361; asc GrainReference=ee0808810738bca; (total 63 bytes);
 2: len 18; hex 53757065727669736f7252656d696e646572; asc SupervisorReminder;;
 3: len 6; hex 0000008b5df4; asc     ] ;;
 4: len 7; hex c7000002850110; asc        ;;
 5: len 5; hex 9997e0f5bb; asc      ;;
 6: len 4; hex 8000ea60; asc    ;;
 7: len 4; hex 3a4f142a; asc :O ;;
 8: len 8; hex 8000000000000000; asc         ;;

*** (2) TRANSACTION:
TRANSACTION 9133557, ACTIVE 0 sec inserting
mysql tables in use 1, locked 1
4 lock struct(s), heap size 1184, 3 row lock(s)
MySQL thread id 30245, OS thread handle 0x7f3b33a4d700, query id 42183671 10.201.200.36 orleans_user update
INSERT INTO OrleansRemindersTable
			(
				ServiceId,
				GrainId,
				ReminderName,
				StartTime,
				Period,
				GrainIdConsistentHash
			)
			VALUES
			(
				_serviceId,
				_grainId,
				_reminderName,
				_startTime,
				_period,
				_grainIdConsistentHash
			)
*** (2) HOLDS THE LOCK(S):
RECORD LOCKS space id 1082 page no 3 n bits 168 index `PRIMARY` of table `orleans`.`OrleansRemindersTable` trx id 9133557 lock_mode X locks gap before rec
Record lock, heap no 70 PHYSICAL RECORD: n_fields 9; compact format; info bits 0
 0: len 30; hex 30303030303030302d303030302d303030302d303030302d303030306235; asc 00000000-0000-0000-0000-0000b5; (total 36 bytes);
 1: len 30; hex 477261696e5265666572656e63653d656530383038383130373338626361; asc GrainReference=ee0808810738bca; (total 63 bytes);
 2: len 18; hex 53757065727669736f7252656d696e646572; asc SupervisorReminder;;
 3: len 6; hex 0000008b5df4; asc     ] ;;
 4: len 7; hex c7000002850110; asc        ;;
 5: len 5; hex 9997e0f5bb; asc      ;;
 6: len 4; hex 8000ea60; asc    ;;
 7: len 4; hex 3a4f142a; asc :O ;;
 8: len 8; hex 8000000000000000; asc         ;;

*** (2) WAITING FOR THIS LOCK TO BE GRANTED:
RECORD LOCKS space id 1082 page no 3 n bits 168 index `PRIMARY` of table `orleans`.`OrleansRemindersTable` trx id 9133557 lock_mode X locks gap before rec insert intention waiting
Record lock, heap no 70 PHYSICAL RECORD: n_fields 9; compact format; info bits 0
 0: len 30; hex 30303030303030302d303030302d303030302d303030302d303030306235; asc 00000000-0000-0000-0000-0000b5; (total 36 bytes);
 1: len 30; hex 477261696e5265666572656e63653d656530383038383130373338626361; asc GrainReference=ee0808810738bca; (total 63 bytes);
 2: len 18; hex 53757065727669736f7252656d696e646572; asc SupervisorReminder;;
 3: len 6; hex 0000008b5df4; asc     ] ;;
 4: len 7; hex c7000002850110; asc        ;;
 5: len 5; hex 9997e0f5bb; asc      ;;
 6: len 4; hex 8000ea60; asc    ;;
 7: len 4; hex 3a4f142a; asc :O ;;
 8: len 8; hex 8000000000000000; asc         ;;

*** WE ROLL BACK TRANSACTION (1)
```

**Comment by veikkoeeva** [comment]
@shayhatsor You may know this already, but a word of caution that locking in various databases is different and I'm sure so is locking different in Oracle and MySQL (which was one of the motivations to put the queries to DB and not to code. For sure in SQL Server this pattern is called *upsert*. That out of way, this is a common problem and SQL Server follows the common pattern that can be seen, for instance, [here](http://weblogs.sqlteam.com/dang/archive/2007/10/28/Conditional-INSERTUPDATE-Race-Condition.aspx) and [here](http://michaeljswart.com/2011/09/mythbusting-concurrent-updateinsert-solutions/). Hmm, I remember for MySQL this pattern is called `INSERT ... ON DUPLICATE KEY UPDATE`. I'm here better on computer in about five hours, but searching with something like "MySQL, upsert duplicate key update" likely yields desired results.

Last time we were discussing I had hackathon and moving, now I'm waiting for components to arrive for my desktop, but I should be better available. :)

**Comment by shayhatsor** [comment]
I'm starting to work on it. it seems to me that:
unconditional upserts should be changed to: http://dev.mysql.com/doc/refman/5.7/en/insert-on-duplicate.html
 
conditional upserts should be changed to something along the lines of:
* http://stackoverflow.com/questions/16588401/conditional-on-duplicate-key-update-update-only-if-certain-condition-is-true
* http://stackoverflow.com/questions/2469457/insert-on-duplicate-key-update-with-where
* https://thewebfellas.com/blog/conditional-duplicate-key-updates-with-mysql

**Comment by shayhatsor** [comment]
a proposed fix is at #1207

**Comment by gabikliot** [comment]
Fixed via #1207.
